kvbc/src/categorization/block_merkle_category.cpp

```diff
@@ -0,0 +1,398 @@
+  auto hasher = Hasher{};
+  auto hashed_key = hasher.digest(key.data(), key.size());
+  if (auto block_id = getLatestVersion(hashed_key)) {
+    return get(hashed_key, *block_id);
+  }
+  return std::nullopt;
+}
+
+std::optional<BlockId> BlockMerkleCategory::getLatestVersion(const std::string& key) const {
+  auto hasher = Hasher{};
+  auto hashed_key = hasher.digest(key.data(), key.size());
+  return getLatestVersion(hashed_key);
+}
+
+std::optional<BlockId> BlockMerkleCategory::getLatestVersion(const Hash& hashed_key) const {
+  const auto serialized = db_->getSlice(BLOCK_MERKLE_LATEST_KEY_VERSION, hashed_key);
+  if (!serialized) {
+    return std::nullopt;
+  }
+  auto version = LatestKeyVersion{};
+  deserialize(*serialized, version);
+  if (version.block_id == 0) {
+    // 0 is a tombstone sentinel
+    return std::nullopt;
+  }
+  return version.block_id;
+}
+
+void BlockMerkleCategory::multiGet(const std::vector<std::string>& keys,
+                                   const std::vector<BlockId>& versions,
+                                   std::vector<std::optional<MerkleValue>>& values) const {
+  ConcordAssertEQ(keys.size(), versions.size());
+  auto versioned_keys = versionedKeys(keys, versions);
+  multiGet(versioned_keys, versions, values);
+}
+
+void BlockMerkleCategory::multiGet(const std::vector<Buffer>& versioned_keys,
+                                   const std::vector<BlockId>& versions,
+                                   std::vector<std::optional<MerkleValue>>& values) const {
+  auto slices = std::vector<::rocksdb::PinnableSlice>{};
+  slices.reserve(versioned_keys.size());
+  auto statuses = std::vector<::rocksdb::Status>{};
+  statuses.reserve(versioned_keys.size());
+
+  db_->multiGet(BLOCK_MERKLE_KEYS_CF, versioned_keys, slices, statuses);
+
+  values.clear();
+  for (auto i = 0ull; i < slices.size(); ++i) {
+    const auto& status = statuses[i];
+    const auto& slice = slices[i];
+    const auto version = versions[i];
+    if (status.ok()) {
+      values.push_back(MerkleValue{{version, slice.ToString()}});
+    } else if (status.IsNotFound()) {
+      values.push_back(std::nullopt);
+    } else {
+      throw std::runtime_error{"BlockMerkleCategory multiGet() failure: " + status.ToString()};
+    }
+  }
+}
+
+void BlockMerkleCategory::multiGetLatestVersion(const std::vector<std::string>& keys,
+                                                std::vector<std::optional<BlockId>>& versions) const {
+  auto hashed_keys = hashedKeys(keys);
+  multiGetLatestVersion(hashed_keys, versions);
+}
+
+void BlockMerkleCategory::multiGetLatestVersion(const std::vector<Hash>& hashed_keys,
+                                                std::vector<std::optional<BlockId>>& versions) const {
+  auto slices = std::vector<::rocksdb::PinnableSlice>{};
+  slices.reserve(hashed_keys.size());
+  auto statuses = std::vector<::rocksdb::Status>{};
+  statuses.reserve(hashed_keys.size());
```

Comment:
Nit: I think the NativeClient will resize the slices and statuses internally, so these might be redundant.

Reply:
Ah you are right. Will remove to get rid of the clutter. Thanks!